packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts

```diff
@@ -0,0 +1,132 @@
+        return {
+            mainContainer: value.includes("mainContainer"),
+            summarizer: value.includes("summarizer"),
+        };
+    },
+};
+
+/**
+ * Error class raised when a SweepReady object is used, indicating a bug in how
+ * references are managed in the container by the application, or a bug in how
+ * GC tracks those references.
+ *
+ * There's a chance for false positives when this error is raised by a Main Container,
+ * since only the Summarizer has the latest truth about unreferenced node tracking
+ */
+export class SweepReadyUsageError extends LoggingError implements IFluidErrorBase {
+    /** This errorType will be in temporary use (until Sweep is fully implemented) so don't add to any errorType type */
+    public errorType: string = "unreferencedObjectUsedAfterGarbageCollected";
+}
+
+/**
+ * This class encapsulates the logic around what to do when a SweepReady object is used.
+ * There are several tactics we plan to use in Dogfood environments to aid diagnosis of these cases:
+ *  - Closing the main container when either the main or summarizer client detects this kind of violation
+ *  - Throttling the frequency of these crashes via a "Blackout Period" per container per device
+ */
+export class SweepReadyUsageDetectionHandler {
+    private readonly localStorage: Pick<Storage, "getItem" | "setItem">;
+
+    constructor(
+        private readonly uniqueContainerKey: string,
+        private readonly mc: MonitoringContext,
+        private readonly closeFn: (error?: ICriticalContainerError) => void,
+        localStorageImpl?: Pick<Storage, "getItem" | "setItem">,
+    ) {
+        const noopStorage = { getItem: () => null, setItem: () => {} };
+        if (localStorageImpl !== undefined) {
+            this.localStorage = localStorageImpl;
+        } else {
+            try {
+                // localStorage is not defined in Node environment so this throws
+                this.localStorage = localStorage ?? noopStorage;
+            } catch (error) {
+                this.localStorage = noopStorage;
+            }
+        }
+
+        if (this.localStorage === noopStorage) {
+            // This means the Blackout Period logic will not work.
+            this.mc.logger.sendTelemetryEvent({ eventName: "SweepReadyUsageDetectionHandlerNoOpStorage" });
+        }
+    }
+
+    /**
+      * If SweepReady Usage Detection is enabled, close the main container.
+      * If the "Blackout Period" is set, don't close the container more than once in that period.
+      *
+      * Once Sweep is fully implemented, this will be removed since the objects will be gone
+      * and errors will arise elsewhere in the runtime
+     */
+    public usageDetectedInMainContainer(errorProps: ITelemetryProperties) {
+        if (!sweepReadyUsageDetectionSetting.read(this.mc.config).mainContainer) {
+            return;
+        }
+
+        const pastClosuresMap: Record<string, { lastCloseTime: number; } | undefined> = (() => {
```

Comment:
I am wondering if you could avoid reading the local storage and set up this state if `blackoutPeriodDaysKey` is undefined and just close the container. If its not, do all of this logic. It will make the code more readable and easier to understand :-).

Reply:
I thought of that but thought the code was more readable this way :P  I'll take a quick look again.

Reply:
Yeah this will work out and read better, you're right - thanks!